### PR TITLE
Filedrop in browser

### PR DESCRIPTION
**Pull request by CarlSchwan**
:heavy_check_mark: Ready for review

Need https://github.com/nextcloud/server/pull/35736

## Todo

- [x] Upload file ui (currently it's a text field)
- [x] Download user public key
- [x] Lock folder
- [x] Download folder metadata info
- [x] Encrypt file
   - [x] Symmetrically Encrypt and decrypt data
- [x] Upload files
- [x] Update folder metadata
   - [x] Asymmetrically encrypt data
   - [x] Add all the other metadata
- [x] Upload new metadata
- [x] Upload folder data metadata
- [x] Unlock folder
- [x] Add HTTP header to mark a request as supporting e2ee
- [x] Make it pretty :)
- [x] ~~Check share for subfolders~~ couldn't reproduce the issue

## Screencast

[Screencast from 2023-01-02 10-03-04.webm](https://user-images.githubusercontent.com/6653109/210211387-8d7b6aea-a321-4c16-ae20-3f8848e84571.webm)

### CODE REVIEW

**Comment by tobiasKaminsky** [comment]
> * Add api for clients to register a file drop

What is this? 
I would assume that clients do not need anything, as thy can work with metadata and files via existing endpoints?

**Comment by CarlSchwan** [comment]
> > * Add api for clients to register a file drop
> 
> What is this? I would assume that clients do not need anything, as thy can work with metadata and files via existing endpoints?

Depends if we want to reuse the normal file drop of not. If we want to reuse it, it makes things more complicated as we need to make also large changes to the server repo to add way to extend the normal file drop.

I was thinking about keeping it simple and just have a special end to end encryption file drop. I didn't implement it yet. What do you think?

**Comment by tobiasKaminsky** [comment]
> I was thinking about keeping it simple and just have a special end to end encryption file drop.

Whatever you prefer :-)
But beware: 
This specific file drop url then must be also shown in /ocs/v2.php/apps/files_sharing/api/v1/shares, as we use this to show all shares / links.

**Comment by tobiasKaminsky** [comment]
@artonge this seems not to work with php 7.4, which we support for NC25 (and older…).

**Comment by artonge** [comment]
> @artonge this seems not to work with php 7.4, which we support for NC25 (and older…).

I adapted the code to php7.4 and fixed some linting errors. Should be better now. But unit tests seem to fail. Use with caution :)

**Comment by tobiasKaminsky** [comment]
Ah, sorry. After my 1:1 we discovered that it is enough to have it it for NC26. Sorry!

**Comment by tobiasKaminsky** [comment]
image

This should still be shown as an encrypted folder, even if it is shared.

**Comment by artonge** [comment]
I update the first comment with a screencast.
@nimishavijay or @jancborchardt could you give some design input ? :)

**Review by jancborchardt — APPROVED** [review]
Very nice @artonge @CarlSchwan! :)

Looks good to go, 2 possible enhancements:
- [x] The scroll container of the file list being limited looks strange and boxed in. Could just scroll the whole page?
- [ ] Can we have the standard filetype icons / file thumbnails here?

It's good to go without the adjustments though, feel free to just open them as follow-up issues. :)

**Comment by artonge** [comment]
> *  The scroll container of the file list being limited looks strange and boxed in. Could just scroll the whole page?

If we allow the whole page to scroll, then the upload button can disappear when there is a large list of uploaded files. Would this be ok ?

>     *  Can we have the standard filetype icons / file thumbnails here?

Good idea, I will look into it

**Comment by jancborchardt** [comment]
> > *  The scroll container of the file list being limited looks strange and boxed in. Could just scroll the whole page?
> 
> If we allow the whole page to scroll, then the upload button can disappear when there is a large list of uploaded files. Would this be ok ?

Yeah good point. The button could be sticky via CSS, then it's best of both worlds?

File: lib/Controller/MetaDataController.php [hunk 1]
```diff
@@ -115,21 +122,23 @@ public function setMetaData(int $id, string $metaData): DataResponse {
 	/**
 	 * Update metadata
 	 *
+	 * @PublicPage
```
Comment (artonge) [inline lib/Controller/MetaDataController.php:125]: Important: the `PUT /.../meta-data/{id}` is now open to public access

File: lib/Controller/LockingController.php [hunk 2]
```diff
@@ -68,24 +73,28 @@ public function __construct(string $AppName,
...
 		$this->l10n = $l10n;
+		$this->shareManager = $shareManager;
 	}
 
 	/**
 	 * Lock folder
 	 *
 	 * @NoAdminRequired
 	 * @E2ERestrictUserAgent
+	 * @PublicPage
```
Comment (artonge) [inline lib/Controller/LockingController.php:84]: Important: the `POST /.../lock/{id}` is now open to public access

File: lib/Controller/LockingController.php [hunk 3]
```diff
@@ -108,18 +117,21 @@ public function lockFolder(int $id): DataResponse {
...
```
Comment (artonge) [inline lib/Controller/LockingController.php:120]: Important: the `DELETE /.../lock/{id}` is now open to public access

File: lib/Controller/MetaDataController.php [hunk 4]
```diff
@@ -44,42 +44,49 @@
...
 		$this->lockManager = $lockManager;
 		$this->l10n = $l10n;
+		$this->shareManager = $shareManager;
 	}
 
 	/**
 	 * Get metadata
 	 *
 	 * @NoAdminRequired
+	 * @PublicPage
```
Comment (artonge) [inline lib/Controller/MetaDataController.php:80]: Important: the `GET /.../meta-data/{id}` is now open to public access

File: lib/Controller/MetaDataController.php (hunk 1 above)
Comment (tobiasKaminsky) [reply lib/Controller/MetaDataController.php:125]: I think this is not good, as then everyone can overwrite the current metadata.
This should be a separate endpoint, which only allows to append filedrop information.
(I discussed it back then with Carl this way…)

**Review by artonge — CHANGES_REQUESTED** [review]
Check if it works for subfolders

**Comment by tobiasKaminsky** [comment]
@artonge please do not forget this PR 💙

**Comment by artonge** [comment]
> @artonge please do not forget this PR blue_heart

@tobiasKaminsky I added a new route to allow updating only the filedtop part of the metadata. Could you check/adapt the clients?

The interface is almost the same as for updating the metadata. Here is en example of adapting js code to the new route:

https://github.com/nextcloud/end_to_end_encryption/pull/367/commits/40ceed7595b580d5994a9dadd7149e8feed28396#diff-398009983aa51c7d24c9ebf499cd645fa2a4ce2ae50d194bc850401a7557a241L83-R85

```
PUT /api/v1/meta-data/{folder-id}/filedrop'
{
    ... your filedrop metadata
}
```

**Comment by artonge** [comment]
Rebased

**Pull request merged**